encoders/firebase-encoders-reflective/src/main/java/com/google/firebase/encoders/reflective/ReflectiveObjectDecoderProvider.java

```diff
@@ -0,0 +1,214 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.encoders.reflective;
+
+import static com.google.firebase.encoders.reflective.ReflectiveDecoderHelper.*;
+
+import androidx.annotation.NonNull;
+import com.google.firebase.decoders.CreationContext;
+import com.google.firebase.decoders.FieldRef;
+import com.google.firebase.decoders.ObjectDecoder;
+import com.google.firebase.decoders.ObjectDecoderContext;
+import com.google.firebase.decoders.TypeCreator;
+import com.google.firebase.decoders.TypeToken;
+import com.google.firebase.encoders.EncodingException;
+import com.google.firebase.encoders.FieldDescriptor;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+import java.util.HashMap;
+import java.util.Map;
+
+class ReflectiveObjectDecoderProvider implements ObjectDecoderProvider {
+
+  static ReflectiveObjectDecoderProvider INSTANCE = new ReflectiveObjectDecoderProvider();
+
+  private ReflectiveObjectDecoderProvider() {}
+
+  @NonNull
+  @Override
+  public <T> ObjectDecoder<T> get(@NonNull Class<T> clazz) {
+    return new ReflectiveObjectDecoderImpl<>();
+  }
+
+  private static class ReflectiveObjectDecoderImpl<T> implements ObjectDecoder<T> {
+
+    private final Map<String, FieldRef<?>> refs = new HashMap<>();
+    private final Map<String, FieldDescriptor> descriptors = new HashMap<>();
+    private final Map<String, ReflectiveFieldSetter> fieldSetters = new HashMap<>();
+
+    private ReflectiveObjectDecoderImpl() {}
+
+    @NonNull
+    @Override
+    public TypeCreator<T> decode(@NonNull ObjectDecoderContext<T> ctx) {
+      Class<T> clazz = ctx.getTypeToken().getRawType();
+      readMethods(clazz);
+      readFields(clazz);
```

Comment:
Will fields overwrite the methods that were just read above?

Reply:
no, it won't. Fields will only be written when their associated setter methods were not found.